Implement show for NLPModels

abelsiqueira opened the pull request.

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105742) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105743) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105744) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105745) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105746) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105747) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105748) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105749) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105750) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105751) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105752) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105753) | 
| SolverTools | [![](https://img.shields.io/badge/master-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105754) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683105755) |

dpo commented on src/nlp_types.jl:
```diff
@@ -4,6 +4,27 @@ export has_bounds, bound_constrained, unconstrained, linearly_constrained,
+  end
+  print(io, output)
+end
+
```
Let's just show full statistics without trying to be too clever. Something like [this](https://github.com/optimizers/NLPLab/blob/master/%2Bmodel/NlpModel.m#L281) or like what Ipopt and Knitro show before a solve.

dpo commented on src/NLSModels.jl:
```diff
@@ -9,6 +9,24 @@ export AbstractNLSModel, nls_meta, NLSCounters, reset!,
 
 abstract type AbstractNLSModel <: AbstractNLPModel end
 
+function show(io :: IO, nls :: AbstractNLSModel)
```
I think this is the `show` for a `NLSModelMeta`.

abelsiqueira replied:
This uses information from `nls.meta` and `nls.nls_meta`. I can redefine `NLSMeta` to keep all `meta` information as well. It would be a huge change, though, but it can be done.

dpo replied:
What I mean is that this method could just call `show(nls.meta)` and also `show(nls.nls_meta)`.

abelsiqueira replied:
By separating, I think I'll have to drop the 3-column format.

dpo commented on src/autodiff_model.jl:
```diff
@@ -49,6 +49,8 @@ mutable struct ADNLPModel <: AbstractNLPModel
   c
 end
 
+show_header(:: ADNLPModel) = "ADNLPModel - Model with automatic differentiation\n"
```
And this is the `show` for a model, which should call `show(meta)`.

abelsiqueira commented:
I forgot to say how it looks:
```
ADNLSModel - Nonlinear least-squares model with automatic differentiation
  Problem name: Generic
  Total variables………2       Total constraints…1       Total residuals………2
    free………………………………2         linear…………………………0         linear…………………………0
    lower……………………………0         nonlinear…………………1         nnzj………………………………4
    upper……………………………0         equality……………………1         nnzj………………………………4
    low/upp………………………0         lower……………………………0         nnzh………………………………3
    fixed……………………………0         upper……………………………0
    nnzh………………………………3         low/upp………………………0
                              nnzj………………………………2
```

abelsiqueira commented:
Btw, should we show the counters in some way?

abelsiqueira commented:
New way models are shown. The last lines are from `nls_meta`.
```
ADNLSModel - Nonlinear least-squares model with automatic differentiation
  Problem name: Generic
  Total variables………2       Total constraints…1
    free………………………………2         linear…………………………0
    lower……………………………0         nonlinear…………………1
    upper……………………………0         equality……………………1
    low/upp………………………0         lower……………………………0
    fixed……………………………0         upper……………………………0
    nnzh………………………………3         low/upp………………………0
                              nnzj………………………………2
  Total residuals………2
    linear…………………………0
    nonlinear…………………2
    nnzj………………………………4
    nnzh………………………………3
```

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642930) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642931) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642932) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642933) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642934) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642935) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642936) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642937) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642938) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642939) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642940) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642941) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642942) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683642943) |

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951875) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951876) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951877) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951878) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951879) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951880) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951881) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951882) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951883) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951884) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951885) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951886) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951887) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/683951888) |

dpo commented:
Many thanks for this. I'm not sure how to put the three on one row, but I feel we should separate the methods from the `meta` objects from those from the models.

Any idea what's failing in the docs build?

dpo commented:
> Btw, should we show the counters in some way?

Yes, I think counters could have their own `show`.

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063218) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063219) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063220) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063221) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063222) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063223) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063224) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063225) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063226) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063227) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063228) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063229) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063230) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684063231) |

abelsiqueira commented:
> Any idea what's failing in the docs build?

abelsiqueira commented:
> Many thanks for this. I'm not sure how to put the three on one row, but I feel we should separate the methods from the `meta` objects from those from the models.

Leaving separated wasn't so ugly. What I don't like is writing `show` for everything. I'm thinking of leaving the `show_header` as well and having the generic `show` call (i) `show_header`, (ii) `show` for `meta` and (iii) show for `counters.

> Any idea what's failing in the docs build?

I made a mistake, will fix soon.

I'll add the counter's show.

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745315) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745316) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745318) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745319) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745320) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745321) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745322) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745324) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745325) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745326) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745327) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745328) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745330) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684745332) |

abelsiqueira commented:
Current outputs:
```
ADNLPModel - Model with automatic differentiation
  Problem name: ex
  Total variables………2       Total constraints…2     
    free………………………………0         linear…………………………0     
    lower……………………………2         nonlinear…………………2     
    upper……………………………0         equality……………………1     
    low/upp………………………0         lower……………………………1     
    fixed……………………………0         upper……………………………0     
    nnzh………………………………3         low/upp………………………0     
                              nnzj………………………………4     
  Counters:
    obj……………2         grad…………0         cons…………0         jcon…………0         jgrad………0     
    jac……………0         jprod………0         jtprod……0         hess…………0         hprod………0     
    jhprod……0  
```
and
```
ADNLSModel - Nonlinear least-squares model with automatic differentiation
  Problem name: Generic
  Total variables………2       Total constraints…1
    free………………………………2         linear…………………………0
    lower……………………………0         nonlinear…………………1
    upper……………………………0         equality……………………0
    low/upp………………………0         lower……………………………1
    fixed……………………………0         upper……………………………0
    nnzh………………………………3         low/upp………………………0
                              nnzj………………………………2
  Total residuals………2
    linear…………………………0
    nonlinear…………………2
    nnzj………………………………4
    nnzh………………………………3
  Counters:
    obj……………………………………0         grad…………………………………0         cons…………………………………0         jcon…………………………………0
    jgrad………………………………0         jac……………………………………0         jprod………………………………0         jtprod……………………………0
    hess…………………………………0         hprod………………………………0         jhprod……………………………0         residual………………………0
    jac_residual……………0         jprod_residual………0         jtprod_residual……0         hess_residual…………0
    jhess_residual………0         hprod_residual………0

```

dpo commented:
Looks great except a blank line missing before the counters of a NLSModels.

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929113) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929114) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929115) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929116) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929117) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929118) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929119) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929120) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929121) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929122) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929123) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929124) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929125) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684929126) |

abelsiqueira commented:
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930594) | [![](https://img.shields.io/badge/v0.8.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930595) | 
| CUTEst | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930596) | [![](https://img.shields.io/badge/v0.10.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930597) | 
| CaNNOLeS | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930598) | [![](https://img.shields.io/badge/v0.2.0-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930599) | 
| NLPModelsIpopt | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930600) | [![](https://img.shields.io/badge/v0.3.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930601) | 
| NLPModelsJuMP | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930602) | [![](https://img.shields.io/badge/v0.6.3-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930603) | 
| QuadraticModels | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930604) | [![](https://img.shields.io/badge/stable-Fail-red)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930605) | 
| SolverTools | [![](https://img.shields.io/badge/master-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930606) | [![](https://img.shields.io/badge/v0.1.12-Pass-green)](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/684930607) |
